### PR TITLE
Add example for KNRM training on WikiQA

### PR DESCRIPTION
**Pull request by hkvision**
RDD based Scala example.
Will add array based API, website doc and Python part in following PRs.

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/Relations.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    val negative = relations.filter(_.label == 0).groupBy(_.text1ID)
+    positive.cogroup(negative).flatMap(x => {
+      val posIDs = x._2._1.flatten.toArray.map(_.text2ID)
+      val negIDs = x._2._2.flatten.toArray.map(_.text2ID)
+      posIDs.flatMap(y => negIDs.map(z => RelationPair(x._1, y, z)))
+    })
+  }
+}
+
+case class Relation(text1ID: String, text2ID: String, label: Int)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/Relations.scala]: make it common utility (id1, id2, label)

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 2]
```diff
@@ -341,13 +481,17 @@ class DistributedTextSet(var rdd: RDD[TextFeature]) extends TextSet {
   }
 
   override def generateWordIndexMap(
-    removeTopN: Int = 0, maxWordsNum: Int = -1): Map[String, Int] = {
+    removeTopN: Int = 0,
+    maxWordsNum: Int = -1,
+    minFreq: Int = 1,
+    existingMap: Map[String, Int] = null): Map[String, Int] = {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: check if removeTopN>0 or minFreq>1 before performing these costly operations (like reduce, sort, etc.)

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/Relations.scala (hunk 1 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/Relations.scala]: Sure.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 2 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: OK. I suppose if any one of `removeTopN` > 0, `maxWordsNum` > 0 (only consider the maximum number of words with highest frequencies) or `minFreq` > 1 is satisfied, we need to do the sort.
If none of the above is satisfied, then we don't need to do the sort, but call `distinct()` to get all words.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/common/Relations.scala [hunk 3]
```diff
@@ -51,16 +51,24 @@ object Relations {
...
+ */
+case class Relation(id1: String, id2: String, label: Int)
 
-case class RelationPair(text1ID: String, text2PosID: String, text2NegID: String)
+/**
+ * It is made up of two relations:
+ * Relation(id1, id2Pos, label>0)
+ * Relation(id1, id2Neg, label=0)
+ */
+case class RelationPair(id1: String, id2Pos: String, id2Neg: String)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/common/Relations.scala]: id2Positive and id2Negtive?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 4]
```diff
@@ -439,13 +446,23 @@ class LocalTextSet(var array: Array[TextFeature]) extends TextSet {
...
+    require(minFreq >= 1, "minFreq should be a positive integer")
+    val words = if (removeTopN == 0 && maxWordsNum == -1 && minFreq == 1) {
+      array.flatMap(_.getTokens).distinct
     }
-    val wordIndex = TextSet.wordsToMap(frequencies, existingMap)
+    else {
+      var frequencies = array.flatMap(_.getTokens)
+        .groupBy(identity).mapValues(_.length).toArray
+      if (minFreq > 1) frequencies = frequencies.filter(_._2 >= minFreq)
+      var res = frequencies.sortBy(- _._2).map(_._1)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: only need to sort if removeTopN > 0 or maxWordsNum > 0, yes?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 5]
```diff
@@ -487,13 +504,23 @@ class DistributedTextSet(var rdd: RDD[TextFeature]) extends TextSet {
...
+      "maxWordsNum should be either -1 or a positive integer")
+    require(minFreq >= 1, "minFreq should be a positive integer")
+    val words = if (removeTopN == 0 && maxWordsNum == -1 && minFreq == 1) {
+      rdd.flatMap(_.getTokens).distinct().collect()
+    }
+    else {
+      var frequencies = rdd.flatMap(_.getTokens)
+        .map(word => (word, 1)).reduceByKey(_ + _)
+      if (minFreq > 1) frequencies = frequencies.filter(_._2 >= minFreq)
+      var res = frequencies.sortBy(- _._2).map(_._1).collect()
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: only need to sort if removeTopN > 0 or maxWordsNum > 0, yes?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala (hunk 4 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: Ar... Yes. Sorry for overlooking this. Updated. Only sort in this case.

**Comment by hkvision** [comment]
Any further comments? @jason-dai @shane-huang

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+## Download Analytics Zoo
+You can download Analytics Zoo prebuilt release and nightly build package from [here](https://analytics-zoo.github.io/master/#release-download/) and extract it.
+
+
+## Data Preparation
+The data used in this example are:
+- [WikiQA](https://www.microsoft.com/en-us/download/details.aspx?id=52419): a new publicly available set of question and sentence pairs.
+- Word Embeddings: We use [`glove.840B.300d.txt`](http://nlp.stanford.edu/data/glove.840B.300d.zip) in this example. You can also choose other available GloVe embeddings from [here](https://nlp.stanford.edu/projects/glove/).
+
+For WikiQA dataset, we refer to MatchZoo to process it into corpus and relations: `question_corpus.csv`, `answer_corpus.csv`, `relation_train.csv` and `relation_valid.csv`.
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: May change `Data preparation` part as below

Data files needed in this example:
- Word Embeddings: We use [`glove.840B.300d.txt`](http://nlp.stanford.edu/data/glove.840B.300d.zip) in this example. You can also choose other available GloVe embeddings from [here](https://nlp.stanford.edu/projects/glove/).
- QA data set : [WikiQA](https://www.microsoft.com/en-us/download/details.aspx?id=52419). We pre-processed the original WikiQA dataset into question/answer corpus and relations files in the same way as MatchZoo does. Thus this example takes below input files instead of the original WikiQA format: 
     - `question_corpus.csv`: the question corpus
     - `answer_corpus.csv`: the answer corpus
     - `relation_train.csv` and `relation_valid.csv`. question and answer correspondence split into train and validation set.

Comment (shane-huang) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: You may need to tell the users where to download these files or how to process wikiqa dataset to obtain them (may an instruction in match-zoo or code or write some)

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    --dataPath ${dataDir} \
+    --embeddingFile ${glovePath}
+```
+See [here](#options) for more configurable options for this example.
+
+
+## Options
+* `--dataPath` This option is __required__. The directory containing the corpus and relations.
+* `--embeddingFile` This option is __required__. The file path to GloVe embeddings.
+* `--text1Length` The length of each question. Default is 10.
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: Since this example is a QA ranker, may be you can change parameters text1xxx an text2xxx to questionxxx and answerxxx.  It's okay to keep KNRM interface to use text1xxx, text2xxx since it's multi-purpose.

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+* `--text2Length` The length of each answer. Default is 40.
+* `--tokenLength` The size of each word vector. GloVe supports tokenLength 50, 100, 200 and 300. Default is 300 for `glove.840B.300d.txt`.
+* `-b` `--batchSize` The number of samples per gradient update. Default is 200.
+* `--nbEpoch` The number of iterations to train the model. Default is 30.
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: add `-e` short option for epochs

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+* `-l` `--learningRate` The learning rate for the KNRM model. Default is 0.001.
+* `--partitionNum` The number of partitions to cut the dataset into. Datault is 4.
+* `--model` Specify this option only if you want to load an existing KNRM model and in this case its path should be provided.
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: add `-m` short option for model

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+
+## Results
+You can find the validation information from the log during the training process:
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: change `log` to `console log`

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 11]
```diff
@@ -241,29 +276,137 @@ object TextSet {
...
-      val textArr = texts.zip(labels).map{case (text, label) =>
-        TextFeature(text, label)
-      }.toArray
-      TextSet.array(textArr)
+      TextSet.array(features.toArray)
     }
     textSet
   }
 
+  // Without header
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: more formal documentation needed for API methods (what the function do , and spec of arguments)

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 12]
```diff
@@ -241,29 +276,137 @@ object TextSet {
...
+      val src = Source.fromFile(path)
+      val textArray = src.getLines().toArray.map(line => {
+        val subs = line.split(",", 2)
+        TextFeature(subs(1), uri = subs(0))
+      })
+      TextSet.array(textArray)
+    }
+  }
+
+  def readParquet(path: String, sqlContext: SQLContext): DistributedTextSet = {
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: more formal documentation needed for API methods (what the function do , and spec of arguments)

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 13]
```diff
@@ -241,29 +276,137 @@ object TextSet {
...
+    val textRDD = sqlContext.read.parquet(path).rdd.map(row => {
+      val uri = row.getAs[String](TextFeature.uri)
+      val text = row.getAs[String](TextFeature.text)
+      TextFeature(text, uri = uri)
+    })
+    TextSet.rdd(textRDD)
+  }
+
+  // Generate RelationPairs: (text1ID, text2PosID, text2NegID)
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: more formal documentation needed for API methods (what the function do , and spec of arguments)

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 14]
```diff
@@ -241,29 +276,137 @@ object TextSet {
...
+      val feature = Tensor(pairedIndices, Array(2, text1.length + text2Pos.length))
+      val label = Tensor(Array(1.0f, 0.0f), Array(2, 1))
+      textFeature(TextFeature.sample) = Sample(feature, label)
+      textFeature
+    })
+    TextSet.rdd(res)
+  }
+
+  // Generate RelationLists: each question together with all its answers and labels
+  // and transform each RelationList to a TextFeature.
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: more formal documentation needed for API methods (what the function do , and spec of arguments)

File: zoo/src/main/scala/com/intel/analytics/zoo/models/textmatching/TextMatcher.scala [hunk 15]
```diff
@@ -35,4 +39,113 @@ abstract class TextMatcher[T: ClassTag](
...
+
+  def evaluateMAP(
+      textSet: TextSet,
+      threshold: Double = 0.0): Double = {
+    val map = evaluate(textSet, TextMatcher.map(threshold))
+    logger.info(s"map: $map")
+    map
+  }
+
+  def evaluateNDCG(
```
Comment (shane-huang) [inline zoo/src/main/scala/com/intel/analytics/zoo/models/textmatching/TextMatcher.scala]: maybe further decouple NDCG and MAP to a util class and mixin it with both Recommender and TextMatcher.

**Comment by shane-huang** [comment]
In general looks good to me. Just a few minor changes needed.

File: pyzoo/test/zoo/models/textmatching/test_knrm.py [hunk 16]
```diff
@@ -55,7 +55,7 @@ def kernel(x):
             KM.append(mm_sum)
 
         Phi = Lambda(lambda x: K.tf.stack(x, 1))(KM)
-        out_ = Dense(1, init="uniform", activation="sigmoid", name="dense")(Phi)
```
Comment (shane-huang) [inline pyzoo/test/zoo/models/textmatching/test_knrm.py]: maybe add options for ranking and classification and use different "output layers" and write in documentation which losses are commended for ranking and classification respectively

**Comment by shane-huang** [comment]
In general looks good to me. Just a few minor changes needed. See my comments above.

File: zoo/src/main/scala/com/intel/analytics/zoo/models/textmatching/TextMatcher.scala (hunk 15 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/models/textmatching/TextMatcher.scala]: Added `Ranker` as a trait for mixin.

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/models/textmatching/TextMatcher.scala]: I just realized that for different models the argument type would be different. For TextMatcher, we expect `TextSet`, while for Recommender, we probably expect `RDD[UserItemFeature]`...
Anyway, I first put this trait here for future flexibility.

File: pyzoo/test/zoo/models/textmatching/test_knrm.py (hunk 16 above)
Comment (hkvision) [reply pyzoo/test/zoo/models/textmatching/test_knrm.py]: Added `targetMode = ranking | classification`. This PR focuses on Scala side only. Will update Python in a following PR very soon.

**Comment by hkvision** [comment]
Jenkins passed http://10.239.47.211:18888/job/ZOO-PR-Validation/1445/

**Comment by hkvision** [comment]
Since it is a big PR, I merge it first and further comments if any can be addressed in following PRs. Thanks for your reviews @jason-dai @shane-huang
